File: 77. Combinations/77.go
package leetcode

func combine(n int, k int) [][]int {
	nums := make([]int, 0, n)
	for i:=1; i<n+1; i++{
		nums = append(nums, i)
	}

	res := make([][]int, 0)

	dfs(k, &res, []int{}, nums)
	return res
}

func dfs(n int, res *[][]int, tmp []int, nums []int){
	if n == 0{
		*res = append(*res, tmp)
	}

	if len(nums)<n{
		return
	}

	for i,v := range nums{
		if len(nums) - i < n{
			break
		}
		newtmp := make([]int, 0, len(tmp)+1)
		newtmp = append(newtmp, tmp...)
		newtmp = append(newtmp, v)

		newnums := make([]int, 0, len(nums)-i-1)
		newnums = append(newnums, nums[i+1:]...)

		dfs(n-1, res, newtmp, newnums)
	}
}

File: 77. Combinations/_77.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        res = list()
        nums = list(range(1, n+1))
        self.dfs(k, res, [], nums)
        return res

    def dfs(self, n, res, tmp, nums):
        if n == 0:
            res.append(tmp)
        
        if len(nums)<n:
            return

        for i,v in enumerate(nums):
            if len(nums)-i < n:
                break
            
            self.dfs(n-1, res, tmp+[v], nums[i+1:])

